### DISCUSSION TITLE
Difficulty of Solo Survival Challenges

### DISCUSSION THREAD

**Discussion by coreyja** [BattlesnakeOfficial/feedback#95] [General]
I'm working towards having the badge for each challenge, and the one that has given me the most trouble by far is the Solo Survival series, especially the last couple. I have two specific 'issues' with this set of challenges:

## Food Spawn on Last Empty Space can ruin a game
1. Food spawns when eating food can 'end' a game, and create a state where you have just eaten food but every space on the board is filled. It's impossible to continue on this board (besides simply filling in the rest of the coordinates). The leads to the 'end game' of this challenge being VERY dependent on the luck of the draw for each game. For instance I probably ran my snake through the last few stages a couple hundred times lol 

## Last Tier Might be Impossible
2. I think the last stage of 2500 _might_ be impossible to actually achieve. Chris said on Discord that they were calculated by figuring out the theoretical max and dividing it in ~half, but I wonder if there was a flaw in that math. I think 2500 is RIGHT on the edge of the theoretical max, maybe just a hair above it. Below I will try to lay out my math, which checks out with how my scores have been for my snake

Givens:
- It takes 300 Turns for the board to fill up with food 
  - This is about right from what I've seen but we could math this out too
- Before the first 300 turns its possible to always avoid food.
  - This Probably isn't always true but we are trying to fix a max, so lets say we can
- You Start at Turn 1 with a size of 3
- Board Size of 7 x 7

Ok so given the first two 'givens' first lets figure out the smallest we could possibly be at turn 300 when the board will be full of food

We start a 3, and can do 100 turns in between eating (assuming we eat on the 100th turn). So on turn 100 we can be length 4, 200 => 4, 300 => 6

Ok so we could be only 6 long by turn 300.

Now the best strategy (I believe) is to chase your tail until your health is at 1, and the eat the closest food. And then continue to chase your tail.
Its not always possible to get to 1 health given the shape of your snake, but lets say you are having the perfect game. When you go to eat food, you MUST eat at least 2 pieces of food. First you will leave your safe 'circle' and get a piece of food. Now your tail will have moved, but it will NOT this next move. So we need to move to get into a spot where we can continue to tail chase. To do this we must eat another piece of food. And then we can continue chasing our tail

So each 100 turns we will need to eat at least 2 food, and therefor grow in length by 2.

Continuing this pattern also means that when we reach length 50 is when the game will end.

50 - 6 = 44 Foods to Go

44 Food = 22 cycles of 100 turns and 2 food => 2200 turns

And add our two 'halves' together and we get **300 + 2200 === 2500**

So it _might_ be possible. I made some assumptions about a perfect game that I'm not confident can actually hold true for the entire game. But given issue 1, its hard to optimize to this level given how much of the late game is dependent on the random number generator

**Comment by InterestingBrainPoops** [comment]
Is it possible to write a script/program that runs the "perfect game", basically each time it needs a food, a food is there, and when it doesnt, there is no food there. This could help figure out the theoretical maximum. It would also cut out the RNG aspect in finding the maximum number of turns.

**Comment by bvanvugt** [comment]
Yeah, these challenges are generally too difficult / RNG dependent. We agree, thanks for raising the issue :)

It's likely we re-introduce them with some new setups/ideas once we have engine support for some new things...

**Comment by coreyja** [comment]
Wondering if a friendly admin could check for me if anyone has been able to complete the `Solo Survival 2500` challenge, in relation to this issue?

Wondering if it truly is impossible, or if I need to figure out a better strategy lol!

**Comment by chris-bsnake** [comment]
No, we haven't had anyone complete that challenge. Like Brad said, at the high end these challenges need a revamp. For a lot of them Brad and I would come up with some linear algorithm for determining these challenge levels that seems to make sense, but don't always live up to the real world execution.

**Reply by coreyja** [reply]
Awesome thanks for checking for me!

Ya that makes sense, thanks for the explanation!

**Reply by bvanvugt** [reply]
Priority here is to build what we're calling "challenge" maps, such that folks can produce their own challenges using a map definitions, since, as it turns out, chris and i suck at thinking up good challenges :)
